processing/src/main/java/org/apache/druid/segment/filter/OrFilter.java

```diff
@@ -71,6 +78,160 @@ public OrFilter(List<Filter> filters)
+            merged.merge(bundle.getIndex().getIndexCapabilities());
+            // union index only bitmaps together; if all sub-filters are 'index only' bundles we will make an index only
+            // bundle ourselves, else we will use this index as a single value matcher
+            if (index == null) {
+              index = bundle.getIndex().getBitmap();
+            } else {
+              index = index.union(bundle.getIndex().getBitmap());
+            }
+          } else {
+            // index and matcher bundles must be handled separately, they will need to be a single value matcher built
+            // by doing an AND operation between the index and the value matcher
+            // (a bundle is basically an AND operation between the index and matcher if the matcher is present)
+            partialIndexBundles.add(convertBundleToMatcherOnlyBundle(bundle, bundleIndex));
+          }
+        }
+      } else {
+        matcherOnlyBundles.add(bundle.getMatcherBundle());
+      }
+    }
+    final long totalBitmapConstructTimeNs = System.nanoTime() - bitmapConstructionStartNs;
+
+
+    // if all the filters are 'index only', we can make an index only bundle
+    if (indexOnlyBundles.size() + emptyCount == filters.size()) {
+      if (index == null || index.isEmpty()) {
+        return FilterBundle.allFalse(
+            totalBitmapConstructTimeNs,
+            columnIndexSelector.getBitmapFactory().makeEmptyImmutableBitmap()
+        );
+      }
+      if (indexOnlyBundles.size() == 1) {
+        return new FilterBundle(
+            new FilterBundle.SimpleIndexBundle(
+                indexOnlyBundles.get(0).getIndexInfo(),
+                index,
+                indexOnlyBundles.get(0).getIndexCapabilities()
+            ),
+            null
+        );
+      }
+      return new FilterBundle(
+          new FilterBundle.SimpleIndexBundle(
+              new FilterBundle.IndexBundleInfo(
+                  () -> "OR",
+                  selectionRowCount,
+                  totalBitmapConstructTimeNs,
+                  indexOnlyBundles.stream().map(FilterBundle.IndexBundle::getIndexInfo).collect(Collectors.toList())
+              ),
+              index,
+              merged
+          ),
+          null
+      );
+    }
+
+    // if not the index only outcome, we build a matcher only bundle from all the matchers
+    final List<FilterBundle.MatcherBundle> allMatcherBundles = Lists.newArrayListWithCapacity(
+        (indexOnlyBundles.isEmpty() ? 0 : 1) + partialIndexBundles.size() + matcherOnlyBundles.size()
+    );
+    if (!indexOnlyBundles.isEmpty()) {
+      // translate the indexOnly bundles into a single matcher
+      allMatcherBundles.add(
+          convertIndexToMatcherBundle(selectionRowCount, indexOnlyBundles, totalBitmapConstructTimeNs, index)
+      );
+    }
+    allMatcherBundles.addAll(partialIndexBundles);
+    allMatcherBundles.addAll(matcherOnlyBundles);
+
+    return new FilterBundle(
+        null,
```

Comment:
in case matcher bundles are present, can we still use the bitmaps? Specifically, instead of the indexBundle being null here, we can use the unioned bitmap index. It will weed out the definite negatives (0's) when filtering, and the caller can use the matcher bundle to test the 1's

Reply:
I realize that it will run into the same problem that the PR is trying to eliminate.

Reply:
this doesn't work for an OR filter because a `FilterBundle` with both an index and matcher is basically an AND operation between the index and matcher, like if the bitmap has the row set the value matcher must also match the actual value at the row. The things that can produce bundles with both indexes and matchers are actual AND filters, and also filters that have `ColumnIndexCapabilities.isExact()` with a false value (none exist today, but planning for the future). I tried to explain this in the javadocs for `FilterBundle`, lmk if it isn't clear enough.

Anyway, because a bundle with both is an AND, an OR filter can only ever produce index only bundles or matcher only bundles, which is why we convert the child 'index only' bundles and 'index and matcher' bundles into matcher bundles with `convertIndexToMatcherBundle` and `convertBundleToMatcherOnlyBundle` respectively that use the indexes for cheaper matchers based on just checking the bitmap for a set bit instead of running the matcher on the row value.

The reason to union the index only bundles together is my gut says that having a single matcher bitmap to iterate would be a bit cheaper when doing the row scan than having a bunch of separate matchers with their own bitmaps to iterate. We can't combine the 'index and matcher' bundles into this because they need to check their own index and matcher before being OR'd with the rest of the matchers.

Reply:
Thanks for the explanation!